lightning/src/blinded_path/message.rs

```diff
@@ -74,16 +68,106 @@ impl Writeable for ForwardTlvs {
 	}
 }
 
+/// Similar to [`ForwardTlvs`], but these TLVs are for the final node.
+pub(crate) struct ReceiveTlvs {
+	pub tlvs: Option<RecipientData>
+}
+
+impl ReceiveTlvs {
+	pub fn new(tlvs: Option<RecipientData>) -> Self {
+		ReceiveTlvs { tlvs }
+	}
+}
+
+/// Represents additional data appended along with the sent reply path.
+///
+/// This data can be utilized by the final recipient for further processing
+/// upon receiving it back.
+#[derive(Clone, Debug)]
+pub enum RecipientData {
+	/// Represents the data specific to [`OffersMessage`]
+	///
+	/// [`OffersMessage`]: crate::onion_message::offers::OffersMessage
+	OffersContext(OffersData),
+
+	/// Represents the data appended with Custom Onion Message.
+	CustomContext(Vec<u8>),
+}
+
+/// Contains the data specific to [`OffersMessage`]
+///
+/// [`OffersMessage`]: crate::onion_message::offers::OffersMessage
+#[derive(Clone, Debug)]
+pub struct OffersData {
+	/// Payment ID of the outbound BOLT12 payment.
+	pub payment_id: Option<PaymentId>
+}
+
+impl RecipientData {
+	/// Creates a new RecipientData::OffersContext instance.
+	pub fn new_for_offers(payment_id: Option<PaymentId>) -> Self {
+		RecipientData::OffersContext(OffersData {
+			payment_id,
+		})
+	}
+
+	/// Creates a new RecipientData::CustomContext instance.
+	pub fn new_for_custom(data: Vec<u8>) -> Self {
+		RecipientData::CustomContext(data)
+	}
+}
+
+impl Writeable for RecipientData {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), io::Error> {
+		match self {
+			RecipientData::OffersContext(offers_data) => {
+				1u8.write(writer)?; // Identifier for OffersContext
+				offers_data.payment_id.write(writer)?; // Write the payment_id
+			}
+			RecipientData::CustomContext(data) => {
+				2u8.write(writer)?; // Identifier for CustomContext
+				data.write(writer)?; // Write the custom data
+			}
+		}
+		Ok(())
+	}
+}
+
+impl Readable for RecipientData {
+	fn read<R: io::Read>(r: &mut R) -> Result<Self, DecodeError> {
+		let type_id = u8::read(r)?;
+		match type_id {
+			1u8 => {
+				let payment_id = Option::<PaymentId>::read(r)?;
+				Ok(RecipientData::new_for_offers(payment_id))
+			}
+			2u8 => {
+				let data = Vec::<u8>::read(r)?;
+				Ok(RecipientData::new_for_custom(data))
+			}
+			_ => Err(DecodeError::InvalidValue),
+		}
+	}
+}
```

Comment:
Could you use `impl_writeable_tlv_based_enum` on `RecipientData` and `impl_writeable_tlv_based` on `OffersData`?

Reply:
Used the macros to simply the code in [8e4e1aa](https://github.com/lightningdevkit/rust-lightning/commit/8e4e1aabadb5fda9f4d0ec2c845263c94c273643)
Thanks a lot for the pointers!